### PR TITLE
add quantity node type

### PR DESCRIPTION
**Pull request by justinwoo**
This is an illustration of https://github.com/cstrahan/tree-sitter-nix/issues/2#issuecomment-477984487

### CODE REVIEW

**Comment by cstrahan** [comment]
I think this is a great idea :+1:. In an expression like `((blah))`, "quantity" might not be the most intuitive name. Perhaps "parens" or "parenthesized"? Typing the latter sounds like a pain in the ass, so I'm leaning towards the former.

**Comment by cstrahan** [comment]
... eh, I guess I don't care too much between "parens" and "parenthesized". The javascript grammar calls this "parenthesized_expression", which is even wordier.

Admittedly, I'm lazy, and "parenthesized" is fairly left hand biased on a QWERTY layout. But it's more in line with how I've mostly avoided abbreviations in the grammar, and the name does a better job implying what it is, whereas "parens" suggests an empty pair of parentheses.

So I think my final inclination is that "parenthesized" would be the better name between the two options.

**Comment by justinwoo** [comment]
it's a quantity as taught in normal schools no? i couldnt think of something else that was actually descriptive otherwise, and "parenthesized" suggests something very vague to me

**Comment by cstrahan** [comment]
> it's a quantity as taught in normal schools no?

I never went to University, so I couldn't say :slightly_smiling_face:.

I did a google search for "ast quantity", "parse tree quantity", etc, and I didn't see any obvious hits. My copy of "Compilers: Principles, Techniques, and Tools (2nd Edition)" mentions "quantity" twice, but in each case, it's used like so:

> In specifying a translator, it is helpful to attach attributes to programming construct, where an attribute is any **quantity** associated with a construct. Since constructs are represented by grammar symbols, the concept of attributes extends to grammar symbols. Examples of attributes include an integer value associated with a terminal nurn representing numbers, and a string associated with a terminal id representing identifiers.

I can't find anything in "Language Implementation Patterns".

Is there something in the literature that you could point me to that refers to an expression being enclosed within parens/braces as a "quantity"?

> "parenthesized" suggests something very vague to me

I think of "parenthesized" as short for "parenthesized expression", which is what this is, fundamentally, right? The existence of these `(`/`)` tokens does _not_ imply that they're syntactically/semantically necessary (as in `(((12345)))`, or `((((f) (1)) (2)) (3))`), so we can't ascribe much more meaning in the name of this production rule than just "there happens to be a '(' on the left, an expression in the middle, and a matching ')' on the right".

**Comment by justinwoo** [comment]
"quantity" is a term used in regular K-12 math education for expressions within parentheses, not a term i have found from any PL literature.

**Comment by cstrahan** [comment]
Oh, ha, _that_ "regular school". That went over my head.

In K-12 math, the objects being expressed are generally numerical values / quantities, so in that context all things contained within a pair of parens are going to be quantities. But that intuition seems like a bit of a stretch here, since we often wish to speak of things other than numerical quantities (or expressions that reduce to numerical quantities), like functions, or strings, etc.

Another way of looking at it: in the math context, the use of "quantity" is analogous to our "expression" -- if we were to write a small grammar for basic mathematical expressions, "quantity" might be an apt name for the initial production rule. But following this correspondence, that would suggest that we'd have (if we replace "quantity" with "expression"):

```javascript
    expression: $ => seq('(', $._expr, ')'),
```

But then that conflicts with our existing "expression" rule. Besides, it conflates expressions with those that are specifically parenthesized.

TL;DR: I'm pretty sure most people will find "parenthesized" more natural, so I'd like to go with that.

---

FWIW, here are uses in existing language grammars:

* JavaScript: [ParenthesizedExpression](https://tc39.github.io/ecma262/#prod-ParenthesizedExpression)
* Python: [Parenthesized forms](https://docs.python.org/2.0/ref/parenthesized.html)
* JSONiq: [Parenthesized expression](http://www.jsoniq.org/docs/JSONiq/html/ch05s02s02.html)
* Swift: [Parenthesized expression](https://docs.swift.org/swift-book/ReferenceManual/Expressions.html#grammar_parenthesized-expression)

**Comment by justinwoo** [comment]
I guess "parenthesized" will have to be used then, since nobody knows what "quantity" is. I won't push any new commits to this branch though since I'm already depending on it, could you pull the changes to a separate branch?
